File: frontend/src/arkham/types/Key.ts
import * as JsonDecoder from 'ts.data.json';
import { ChaosToken, chaosTokenDecoder } from '@/arkham/types/ChaosToken';

export function keyToId(key: ArkhamKey): string {
  if (key.tag === "TokenKey") {
    return key.contents.id
  }

  return key.tag
}

export type ArkhamKey
  = { tag: "TokenKey", contents: ChaosToken }
  | { tag: "BlueKey" }
  | { tag: "GreenKey" }
  | { tag: "RedKey" }
  | { tag: "YellowKey" }
  | { tag: "PurpleKey" }
  | { tag: "BlackKey" }
  | { tag: "WhiteKey" }
  | { tag: "UnrevealedKey" }

export const arkhamKeyDecoder = JsonDecoder.oneOf<ArkhamKey>([
  JsonDecoder.object({tag: JsonDecoder.literal("TokenKey"), contents: chaosTokenDecoder }, 'tokenKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("BlueKey") }, 'BlueKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("GreenKey") }, 'GreenKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("RedKey") }, 'RedKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("YellowKey") }, 'YellowKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("PurpleKey") }, 'PurpleKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("BlackKey") }, 'BlackKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("WhiteKey") }, 'WhiteKey'),
  JsonDecoder.object({tag: JsonDecoder.literal("UnrevealedKey") }, 'UnrevealedKey'),
], 'ArkhamKey');

File: backend/arkham-api/library/Arkham/Key.hs
{-# LANGUAGE TemplateHaskell #-}

module Arkham.Key where

import Arkham.ChaosToken.Types
import Arkham.Prelude
import Data.Aeson.TH
import Data.UUID (fromWords64)

data ArkhamKey
  = TokenKey ChaosToken
  | RedKey
  | BlueKey
  | GreenKey
  | YellowKey
  | PurpleKey
  | BlackKey
  | WhiteKey
  | UnrevealedKey ArkhamKey
  deriving stock (Show, Eq, Ord, Generic, Data)
  deriving anyclass (ToJSON, ToJSONKey)

class IsKey a where
  toKey :: a -> ArkhamKey

instance IsKey ChaosToken where
  toKey = TokenKey

instance IsKey ArkhamKey where
  toKey = id

keyName :: ArkhamKey -> Text
keyName = \case
  TokenKey token -> chaosTokenLabel token.face
  RedKey -> "Red"
  BlueKey -> "Blue"
  GreenKey -> "Green"
  YellowKey -> "Yellow"
  PurpleKey -> "Purple"
  BlackKey -> "Black"
  WhiteKey -> "White"
  UnrevealedKey _ -> "Unrevealed"

mconcat
  [ makePrisms ''ArkhamKey
  , [d|
      instance FromJSON ArkhamKey where
        parseJSON o = flip (withObject "ArkhamKey") o \v -> do
          tag :: Text <- v .: "tag"
          case tag of
            "SkullKey" ->
              pure
                $ TokenKey
                $ ChaosToken
                  { chaosTokenId = coerce $ fromWords64 6128981282234515924 12039885860129472512
                  , chaosTokenFace = #skull
                  , chaosTokenRevealedBy = Nothing
                  , chaosTokenCancelled = False
                  , chaosTokenSealed = False
                  }
            "CultistKey" ->
              pure
                $ TokenKey
                $ ChaosToken
                  { chaosTokenId = coerce $ fromWords64 9237114041627526267 13667317789959208286
                  , chaosTokenFace = #cultist
                  , chaosTokenRevealedBy = Nothing
                  , chaosTokenCancelled = False
                  , chaosTokenSealed = False
                  }
            "TabletKey" ->
              pure
                $ TokenKey
                $ ChaosToken
                  { chaosTokenId = coerce $ fromWords64 17367115229369288362 9986755224281741872
                  , chaosTokenFace = #tablet
                  , chaosTokenRevealedBy = Nothing
                  , chaosTokenCancelled = False
                  , chaosTokenSealed = False
                  }
            "ElderThingKey" ->
              pure
                $ TokenKey
                $ ChaosToken
                  { chaosTokenId = coerce $ fromWords64 17260509442849393188 12601461981560257361
                  , chaosTokenFace = #elderthing
                  , chaosTokenRevealedBy = Nothing
                  , chaosTokenCancelled = False
                  , chaosTokenSealed = False
                  }
            _ -> $(mkParseJSON defaultOptions ''ArkhamKey) o
      |]
  ]

instance FromJSONKey ArkhamKey
